components/scream/CMakeLists.txt

```diff
@@ -27,8 +27,38 @@ endfunction ()
 enable_testing()
 include(CTest)
 
+# Bring in Kokkos
+set(SCREAM_KOKKOS_CONFIG --with-openmp --with-options=aggressive_vectorization CACHE STRING "How to configure kokkos (generate_makefile args)")
+
+set(KOKKOS_SRC    ${CMAKE_SOURCE_DIR}/../../externals/kokkos)
+set(KOKKOS_BINARY ${CMAKE_BINARY_DIR}/kokkos/install)
+
+if (NOT EXISTS ${KOKKOS_BINARY}/kokkos_generated_settings.cmake)
```

Comment:
This does not allow one to change kokkos settings and trigger a kokkos build. Granted, does not happen often, but would be nice to make it happen. One way could be to have two variables, `SCREAM_KOKKOS_CONFIG` and `SCREAM_KOKKOS_CONFIG_OLD`, the latter being marked as internal and init-ed to empty; in addition to check for the existence of `kokkos_generated_settings.cmake`, we could check whether the `SCREAM_KOKKOS_CONFIG` and `SCREAM_KOKKOS_CONFIG_OLD` variables match. If they don't, a re-build is needed.

Reply:
This is a good point.

I like the idea of storing old Kokkos config state, but a concern is it depends on keeping state in the CMakeCache.txt, which tends to get zorched a lot. Every time the CMakeCache.txt is constructed from scratch, a Kokkos build would be triggered, right?

Another approach would be to store the Kokkos config line in a text file and diff the user-provided line against it. Probably that's a terrible kludge; not sure.

Perhaps another approach is more in the spirit of CMake: Right now, this PR is based on using Kokkos's Makefile-based build. I speculate using Kokkos's CMake-based build might help. In that case, I speculated we'd trigger a partial or full rebuild only if Kokkos options truly change (for the same reason that, in general, zorching a CMakeCache.txt file doesn't trigger a full rebuild of a project; only using new config options might).

Reply:
I think the cmake variable and the stored string are fairly similar. I'd rather rely on cmake cache rather than an external text file. And yes, for each new build (where no cache exists), a new kokkos build should be triggered. But that should already happen, since there should not be any kokkos_generated_settings.cmake file.

As for generate_makefile vs cmake, I also thought about it. Ideally, the `ExternalProject_Add` feature should be used, though it has the disadvantage of not having the kokkos cmake setting generated at configure time (the external project is configured and build at build time, sigh). A more robust way is probably to simply add the kokkos subdirectory, just like any other project. Homme already offers this capability for blas (though I never used it). We could ask for a kokkos dir, and, if not provided, simply add the external/kokkos as a subdirectory. The drawback is that the user should provide all the `KOKKOS_BLAHBLAH` variables to cmake, though I think that would be not too bad.